### PR TITLE
cgoban: init at 3.5.23

### PR DESCRIPTION
**Pull request by savannidgerinel**
###### Motivation for this change

KGS is one of a few popular Go servers, and cgoban is their official application for communicating with it. While it is a simple Java file and easy to download, it would be convenient to just have it as a derivation without any need to know that it's a Java file.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 130285` at 868e9c40 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>cgoban</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/games/cgoban/default.nix:36:3:
```
   |
36 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/games/cgoban/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{ lib
+, stdenv
+, writers
+, fetchurl
+, adoptopenjdk-jre-bin
+, ver ? "3.5.23"
+, hash ? "0srw1hqr9prgr9dagfbh2j6p9ivaj40kdpyhs6zjkg7lhnnrrrcv" }:
+let
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:8]:
```suggestion
}:
let
```

File: pkgs/games/cgoban/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    installPhase = ''
+      mkdir -p $out/lib
+      cp ${src} $out/lib/cgoban.jar
+    '';
+  };
+  cgoban_sh = writers.writeBash "cgoban" "${adoptopenjdk-jre-bin}/bin/java -jar ${cgoban_jar}/lib/cgoban.jar";
+
+in stdenv.mkDerivation {
+  pname = "cgoban";
+  version = ver;
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:32]:
```suggestion
  version = "3.5.23";
```

File: pkgs/games/cgoban/default.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+, ver ? "3.5.23"
+, hash ? "0srw1hqr9prgr9dagfbh2j6p9ivaj40kdpyhs6zjkg7lhnnrrrcv" }:
+let
+  cgoban_jar = stdenv.mkDerivation rec {
+    pname = "cgoban_jar";
+    version = ver;
+
+    src = fetchurl {
+      url = "http://files.gokgs.com/javaBin/cgoban.jar";
+      hash = "sha256:${hash}";
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:15]:
```suggestion
      sha256 = "0srw1hqr9prgr9dagfbh2j6p9ivaj40kdpyhs6zjkg7lhnnrrrcv" ;
```

File: pkgs/games/cgoban/default.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+, stdenv
+, writers
+, fetchurl
+, adoptopenjdk-jre-bin
+, ver ? "3.5.23"
+, hash ? "0srw1hqr9prgr9dagfbh2j6p9ivaj40kdpyhs6zjkg7lhnnrrrcv" }:
+let
+  cgoban_jar = stdenv.mkDerivation rec {
+    pname = "cgoban_jar";
+    version = ver;
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:11]:
```suggestion
    version = "3.5.23";
```

File: pkgs/games/cgoban/default.nix [hunk 5]
```diff
@@ -0,0 +1,48 @@
+{ lib
+, stdenv
+, writers
+, fetchurl
+, adoptopenjdk-jre-bin
+, ver ? "3.5.23"
+, hash ? "0srw1hqr9prgr9dagfbh2j6p9ivaj40kdpyhs6zjkg7lhnnrrrcv" }:
+let
+  cgoban_jar = stdenv.mkDerivation rec {
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:9]: Why do we need this double mkDerivation? Can't we just throw them together?

Comment (savannidgerinel) [reply pkgs/games/cgoban/default.nix:9]: Without the double derivation i"m having a very hard time with the launcher script. `writeBash` doesn't have the path to `$out` at the time that I define it. I can investigate whether I can make a launcher script like this with `makeWrapper`.

Comment (SuperSandro2000) [reply pkgs/games/cgoban/default.nix]: You can get the value of out with ``placeholder "out"``.

File: pkgs/games/cgoban/default.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+, makeWrapper
+}:
+stdenv.mkDerivation rec {
+  pname = "cgoban";
+  version = "3.5.23";
+
+  nativeBuildInputs = [ adoptopenjdk-jre-bin makeWrapper ];
+
+  src = fetchurl {
+    url = "http://files.gokgs.com/javaBin/cgoban.jar";
```
Comment (SuperSandro2000) [inline pkgs/games/cgoban/default.nix:15]: Do we have a versioned download URL available? If not please upload it to web.archive.org.

Comment (savannidgerinel) [reply pkgs/games/cgoban/default.nix:15]: It does not. It does exist already on `web.archive.org`, and I had wayback recrawl the file (http://web.archive.org/web/20210717230228/https://files.gokgs.com/javaBin/cgoban.jar). How do I find the programmatic URL for downloading it?

Comment (SuperSandro2000) [reply pkgs/games/cgoban/default.nix:15]:
```suggestion
    url = "http://web.archive.org/web/20210717230228/https://files.gokgs.com/javaBin/cgoban.jar";
```

you just use this URL. There is no real way to automatically update it.

**Pull request merged**